android/src/main/java/com/robinpowered/Frequency/FrequencyModule.java

```diff
@@ -0,0 +1,141 @@
+package com.robinpowered.Frequency;
+
+import com.facebook.react.bridge.ReactApplicationContext;
+import com.facebook.react.bridge.ReactContextBaseJavaModule;
+import com.facebook.react.bridge.ReactMethod;
+import com.facebook.react.bridge.Promise;
+
+import android.media.AudioFormat;
+import android.media.AudioManager;
+import android.media.AudioTrack;
+import android.os.Build;
+
+public class FrequencyModule extends ReactContextBaseJavaModule {
+    private static final String MODULE_NAME = "Frequency";
+    private static final int SAMPLE_RATE = 44100;
+
+    private AudioSession audioSession;
+
+    public FrequencyModule(ReactApplicationContext reactContext) {
+        super(reactContext);
+    }
+
+    @Override
+    public String getName() {
+        return MODULE_NAME;
+    }
+
+    private void cancel() {
+        audioSession.cancel();
+        audioSession = null;
+    }
+
+    private void complete() {
+        audioSession.complete();
+        audioSession = null;
+    }
+
+    @ReactMethod
+    public void playFrequency(double frequency, double duration, final Promise promise) {
+        if (audioSession != null && audioSession.isPlaying()) {
+            cancel();
+        }
+
+        final int dur = (int) duration;
+        final int numOfSamples = dur * SAMPLE_RATE;
+
+        final double sample[] = new double[numOfSamples];
+        final byte soundData[] = new byte[2 * numOfSamples];
+
+        for (int i = 0; i < numOfSamples; ++i) {
+            sample[i] = Math.sin(2 * Math.PI * i / (SAMPLE_RATE/frequency));
+        }
+
+        // convert to 16 bit pcm sound array
+        // assumes the sample buffer is normalised.
+        int idx = 0;
+        for (double dVal : sample) {
```

Comment:
There's been feedback left a few times now on moving this logic into a separate method. Currently this react method is monolithic with a number of responsibilities. See previous comments